tsfc/coffee_mode.py

```diff
@@ -0,0 +1,294 @@
+                             collect_monomials)
+from gem.node import traversal
+from gem.gem import (Product, Sum, Comparison, Conditional, Division, Indexed,
+                     IndexSum, MathFunction, Power, Failure, one, index_sum,
+                     Terminal, ListTensor, FlexiblyIndexed, LogicalAnd,
+                     LogicalNot, LogicalOr)
+from gem.utils import groupby
+
+
+import tsfc.vanilla as vanilla
+
+flatten = vanilla.flatten
+
+finalise_options = dict(replace_delta=False, remove_componenttensors=False)
+
+
+def Integrals(expressions, quadrature_multiindex, argument_multiindices, parameters):
+    """Constructs an integral representation for each GEM integrand
+    expression.
+
+    :arg expressions: integrand multiplied with quadrature weight;
+                      multi-root GEM expression DAG
+    :arg quadrature_multiindex: quadrature multiindex (tuple)
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argument
+    :arg parameters: parameters dictionary
+
+    :returns: list of integral representations
+    """
+    # Unroll
+    max_extent = parameters["unroll_indexsum"]
+    if max_extent:
+        def predicate(index):
+            return index.extent <= max_extent
+        expressions = unroll_indexsum(expressions, predicate=predicate)
+    # Choose GEM expression as the integral representation
+    expressions = [index_sum(e, quadrature_multiindex) for e in expressions]
+    expressions = replace_delta(expressions)
+    expressions = remove_componenttensors(expressions)
+    expressions = replace_division(expressions)
+    return optimise_expressions(expressions, argument_multiindices)
+
+
+def optimise_expressions(expressions, argument_multiindices):
+    """Perform loop optimisations on GEM DAGs
+
+    :arg expressions: list of GEM DAGs
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argumen
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+
+    def classify(argument_indices, expression):
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, (Indexed, Conditional)):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+
+    argument_indices = tuple(itertools.chain(*argument_multiindices))
+    # Apply argument factorisation unconditionally
+    classifier = partial(classify, set(argument_indices))
+    monomial_sums = collect_monomials(expressions, classifier)
+    monomial_sums = [optimise_monomial_sum(ms, argument_indices) for ms in monomial_sums]
+    return list(map(monomial_sum_to_expression, monomial_sums))
```

Comment:
So we are in mode 2 here (no pre-evaluation).
`sharing_graph_rewrite` does factorisation and hoisting (by calling `factorise` and `licm` with various arguments). There is a `unpick` function call before these, which does inlining of temporaries (which is like expansion of products in GEM) based on a cost model.
In this branch I expand products unconditionally, followed by factorisation. (Hoisting happens automatically in tsfc codegen, so there is no licm pass) In `tempoary_graph` branch, I'm rewriting the cost model for expansion used in `unpick`. It's not in a finished state yet, but it gets hozepfel down to 19000, which is marginally better than coffee level 2 currently.
preevaluation decision is made in `dissect`, which we don't have yet.